File: lance-py/src/structor_lance/sync.py
"""One replica per target: five Lance tables mirroring PocketBase's, pulled
through the records API in ``(stamp, id)`` order and upserted by ``id``.
PocketBase stays the source of truth; this side only ever catches up.

Cursors: append-only tables (events, import_runs) page by ``created``;
rewritten ones (projects, sessions, session_weeks) page by ``updated`` so a
changed row comes around again. PocketBase ids are random, so every run
starts ``REWIND_MS`` before the saved cursor (upserts make the overlap free)
to cover a row committed in the same millisecond with a lower id.
"""

from __future__ import annotations

import copy
import json
import os
import tempfile
import threading
import time
from collections.abc import Callable
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any

import lancedb
from lancedb.index import FTS

from .pb import PB, Cursor
from .schema import TABLES, Table, from_record
from .targets import Target

REWIND_MS = 2000
OPTIMIZE_EVERY_S = 300
PRUNE_AFTER = timedelta(hours=1)  # versions older than this are dropped; the 7-day default is for readers that pin them


def table_names(db: lancedb.DBConnection) -> set[str]:
    """Names in the database. ``list_tables()`` (0.38) returns a page object with ``.tables``; older ``table_names()`` a list."""
    if hasattr(db, "list_tables"):
        page = db.list_tables()
        names = getattr(page, "tables", page)
        return {str(n) for n in names}
    return set(db.table_names())


STAMP_FORMATS = ("%Y-%m-%d %H:%M:%S.%fZ", "%Y-%m-%d %H:%M:%SZ", "%Y-%m-%d %H:%M:%S.%f", "%Y-%m-%d %H:%M:%S")


def stamp_to_dt(stamp: str) -> datetime | None:
    """'2026-09-09 15:00:00.100Z' (or without the fraction / the Z) → aware datetime, or None."""
    s = stamp.strip().replace("T", " ")
    for fmt in STAMP_FORMATS:
        try:
            return datetime.strptime(s, fmt).replace(tzinfo=UTC)
        except ValueError:
            continue
    return None


def dt_to_stamp(dt: datetime) -> str:
    return dt.astimezone(UTC).strftime("%Y-%m-%d %H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"


def rewind(c: Cursor, ms: int) -> Cursor:
    """Cursor ``ms`` earlier than ``c``, id cleared so every row at that stamp qualifies."""
    dt = stamp_to_dt(c.stamp)
    if dt is None:
        return c
    return Cursor(dt_to_stamp(dt - timedelta(milliseconds=ms)), "")


def cmp_cursor(a: Cursor, b: Cursor) -> int:
    if a.stamp != b.stamp:
        return -1 if a.stamp < b.stamp else 1
    return 0 if a.id == b.id else (-1 if a.id < b.id else 1)


class Replica:
    """A target's Lance directory, its sync state, and the pull loop."""

    def __init__(self, target: Target, data_root: Path):
        self.target = target
        self.dir = Path(data_root) / target.name
        self.dir.mkdir(parents=True, exist_ok=True)
        self.pb = PB(target.url, target.email, target.password)
        self.state: dict[str, Any] = self._load_state()
        self._db: lancedb.DBConnection | None = None
        self._run_lock = threading.Lock()
        self._state_lock = threading.RLock()  # the follow thread writes state, request threads read it
        self._wake = threading.Event()
        self._stop = threading.Event()
        self._optimize_due: dict[str, float] = {}

    # ---- state ------------------------------------------------------------

    @property
    def state_path(self) -> Path:
        return self.dir / "sync.json"

    def _load_state(self) -> dict[str, Any]:
        empty = {"target": self.target.name, "url": self.target.url, "tables": {}, "lastRun": "", "lastError": "", "lastDurationMs": 0}
        if not self.state_path.exists():
            return empty
        try:
            return {**empty, **json.loads(self.state_path.read_text())}
        except (OSError, ValueError):
            return empty

    def save_state(self) -> None:
        """Atomic write of sync.json under the state lock. A failure is recorded, never raised."""
        with self._state_lock:
            try:
                fd, tmp = tempfile.mkstemp(prefix="sync.", suffix=".json.tmp", dir=self.dir)
                with os.fdopen(fd, "w") as f:
                    json.dump(self.state, f, indent=2)
                os.replace(tmp, self.state_path)
            except OSError as e:
                self.state["lastError"] = f"state: {e}"[:500]

    def state_copy(self) -> dict[str, Any]:
        """A snapshot for readers on other threads (the admin's /api/status)."""
        with self._state_lock:
            return copy.deepcopy(self.state)

    def mark_fts_built(self, table: str) -> None:
        with self._state_lock:
            st = self.state["tables"].setdefault(table, {"cursor": None, "rows": 0, "lastPull": ""})
            st["ftsBuilt"] = True
            self.save_state()

    # ---- tables -----------------------------------------------------------

    def db(self) -> lancedb.DBConnection:
        if self._db is None:
            self._db = lancedb.connect(str(self.dir))
        return self._db

    def table(self, model: type[Table]) -> lancedb.table.Table:
        db = self.db()
        if model.__table__ in table_names(db):
            return db.open_table(model.__table__)
        return db.create_table(model.__table__, schema=model, exist_ok=True)

    # ---- sync -------------------------------------------------------------

    def sync_now(self) -> int:
        """Pull everything new for every table; returns rows upserted. Serialised."""
        with self._run_lock:
            t0 = time.time()
            total = 0
            try:
                for model in TABLES:
                    total += self._sync_table(model)
                with self._state_lock:
                    self.state["lastError"] = ""
            except Exception as e:  # noqa: BLE001 — the loop must survive anything
                with self._state_lock:
                    self.state["lastError"] = str(e)[:500]
            with self._state_lock:
                self.state["lastRun"] = datetime.now(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")
                self.state["lastDurationMs"] = int((time.time() - t0) * 1000)
                self.save_state()
            return total

    def _sync_table(self, model: type[Table]) -> int:
        tbl = self.table(model)
        with self._state_lock:
            st = self.state["tables"].setdefault(model.__table__, {"cursor": None, "rows": 0, "lastPull": ""})
            saved = Cursor(**st["cursor"]) if st.get("cursor") else None
        cursor = rewind(saved, REWIND_MS) if saved else None
        pulled = 0
        while True:
            items = self.pb.page_after(model.__table__, model.__stamp__, cursor)
            if not items:
                break
            rows = [from_record(model, r) for r in items]
            tbl.merge_insert("id").when_matched_update_all().when_not_matched_insert_all().execute(rows)
            last = items[-1]
            cursor = Cursor(str(last.get(model.__stamp__) or last.get("created", "")), str(last["id"]))
            with self._state_lock:
                if saved is None or cmp_cursor(cursor, saved) > 0:  # never move the saved cursor backwards
                    saved = cursor
                    st["cursor"] = {"stamp": cursor.stamp, "id": cursor.id}
                st["rows"] = int(st.get("rows", 0)) + len(rows)
                st["lastPull"] = datetime.now(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")
                self.save_state()
            pulled += len(rows)
            if len(items) < 1000:
                break
        if model.__fts__ and pulled > 0:
            self._ensure_fts(model, tbl, st)
        if pulled > 0:
            self._maybe_optimize(model, tbl)
        return pulled

    def _maybe_optimize(self, model: type[Table], tbl: Any) -> None:
        """Compact fragments, fold new rows into the FTS index, prune old versions — every table, every few minutes.

        Every merge page is a new version and a new fragment; left alone a busy
        day is thousands of versions and tens of gigabytes.
        """
        due = self._optimize_due.get(model.__table__, 0.0)
        if time.time() < due:
            return
        self._optimize_due[model.__table__] = time.time() + OPTIMIZE_EVERY_S
        tbl.optimize(cleanup_older_than=PRUNE_AFTER)

    def _ensure_fts(self, model: type[Table], tbl: Any, st: dict) -> None:
        if not st.get("ftsBuilt"):
            if tbl.count_rows() == 0:
                return
            tbl.create_index(model.__fts__, config=FTS(), replace=True)
            with self._state_lock:
                st["ftsBuilt"] = True
        # new rows are folded into the index by _maybe_optimize()

    # ---- follow -----------------------------------------------------------

    def wake(self) -> None:
        self._wake.set()

    def stop(self) -> None:
        self._stop.set()
        self._wake.set()

    def follow(self, interval_s: float, log: Callable[[str], None]) -> None:
        """Sync now, then again on every live message or every ``interval_s``. Blocks; run in a thread."""
        threading.Thread(target=self._live_loop, args=(log,), name=f"live-{self.target.name}", daemon=True).start()
        last_logged = ""
        while not self._stop.is_set():
            try:
                n = self.sync_now()
                if n > 0:
                    log(f"{self.target.name}: +{n} rows")
                elif self.state["lastError"] and self.state["lastError"] != last_logged:
                    log(f"{self.target.name}: {self.state['lastError']}")
                last_logged = self.state["lastError"]
            except Exception as e:  # noqa: BLE001
                log(f"{self.target.name}: {e}")
            self._wake.wait(interval_s)
            self._wake.clear()
            time.sleep(1.5)  # let a burst of ingests coalesce into one pull

    def _live_loop(self, log: Callable[[str], None]) -> None:
        while not self._stop.is_set():
            try:
                self.pb.live("structor/live", lambda _msg: self.wake(), self._stop)
            except Exception as e:  # noqa: BLE001
                log(f"live feed {self.target.name}: {e}")
            self._stop.wait(5)

    # ---- lag --------------------------------------------------------------

    def lag(self) -> dict[str, dict[str, int]]:
        """Rows on the PocketBase side minus rows here, per table."""
        out: dict[str, dict[str, int]] = {}
        for model in TABLES:
            try:
                remote = self.pb.count(model.__table__)
            except Exception:  # noqa: BLE001
                remote = -1
            try:
                local = self.table(model).count_rows()
            except Exception:  # noqa: BLE001
                local = -1
            out[model.__table__] = {"remote": remote, "local": local}
        return out

File: lance/src/sync.ts
// One replica per target: three Lance tables mirroring PocketBase's
// projects / sessions / events, pulled through the records API in
// (stamp, id) order and upserted by id. PocketBase stays the source of
// truth; this side only ever catches up.
//
// Cursors: events are immutable, so `created` orders them. sessions and
// projects are rewritten on every ingest (tail state, cwd reconcile), so
// they page by `updated` and a changed row comes around again.

import * as lancedb from "@lancedb/lancedb";
import { Index, makeArrowTable } from "@lancedb/lancedb";
import { Schema, Field, Utf8, Float64, Bool } from "apache-arrow";
import { existsSync, mkdirSync, readFileSync, writeFileSync, renameSync } from "node:fs";
import { join } from "node:path";
import { PB, type Cursor, type PBRecord } from "./pb.ts";
import type { Target } from "./targets.ts";

export interface TableSpec {
  name: "projects" | "sessions" | "events" | "session_weeks" | "import_runs";
  stamp: "created" | "updated";
  schema: Schema;
  fts?: string; // column that gets a full-text index
}

const utf8 = (n: string) => new Field(n, new Utf8(), true);
const num = (n: string) => new Field(n, new Float64(), true);
const bool = (n: string) => new Field(n, new Bool(), true);

export const TABLES: TableSpec[] = [
  {
    name: "projects",
    stamp: "updated",
    schema: new Schema([utf8("id"), utf8("path"), utf8("name"), utf8("encoded_dir"), utf8("cwd"), utf8("host"), utf8("created"), utf8("updated")]),
  },
  {
    name: "sessions",
    stamp: "updated",
    schema: new Schema([
      utf8("id"), utf8("session_id"), utf8("project"), utf8("file_path"), utf8("tier"),
      num("byte_offset"), num("file_size"), num("file_mtime"), num("lines_seen"), num("event_count"),
      utf8("first_ts"), utf8("last_ts"), utf8("first_prompt"), utf8("git_branch"), utf8("cwd"), utf8("model"),
      utf8("created"), utf8("updated"),
    ]),
  },
  {
    name: "events",
    stamp: "created",
    fts: "text",
    schema: new Schema([
      utf8("id"), utf8("session"), utf8("uuid"), utf8("parent_uuid"), utf8("type"), utf8("role"),
      utf8("ts"), utf8("iso_week"), utf8("text"), utf8("tools"), utf8("model"), bool("sidechain"),
      num("line_no"), num("raw_bytes"), utf8("created"),
    ]),
  },
  {
    // one row per (session, ISO week): the ledger the old console's status strip and Weeks view read
    name: "session_weeks",
    stamp: "updated",
    schema: new Schema([
      utf8("id"), utf8("session"), utf8("project"), utf8("iso_week"),
      num("event_count"), num("user_count"), num("assistant_count"), num("tool_count"),
      utf8("first_ts"), utf8("last_ts"), utf8("created"), utf8("updated"),
    ]),
  },
  {
    // one row per ingest call: the Intake ledger
    name: "import_runs",
    stamp: "created",
    schema: new Schema([
      utf8("id"), utf8("session"), utf8("project"), num("from_offset"), num("to_offset"),
      num("lines"), num("inserted"), num("skipped"), utf8("host"), utf8("writer"), utf8("created"),
    ]),
  },
];

export interface TableState {
  cursor: Cursor | null;
  rows: number;        // rows pulled since the replica was created
  lastPull: string;    // ISO time of the last page that returned rows
  ftsBuilt?: boolean;
}

export interface SyncState {
  target: string;
  url: string;
  tables: Record<string, TableState>;
  lastRun: string;
  lastError: string;
  lastDurationMs: number;
}

export class Replica {
  readonly dir: string;
  readonly pb: PB;
  state: SyncState;
  private db?: lancedb.Connection;
  private running: Promise<number> | null = null;
  private wake: (() => void) | null = null;
  private stopped = false;
  private optimizeDueFor = new Map<string, number>();

  constructor(readonly target: Target, dataRoot: string) {
    this.dir = join(dataRoot, target.name);
    mkdirSync(this.dir, { recursive: true });
    this.pb = new PB(target.url, target.email, target.password);
    this.state = this.loadState();
  }

  private get statePath() { return join(this.dir, "sync.json"); }

  private loadState(): SyncState {
    const empty: SyncState = { target: this.target.name, url: this.target.url, tables: {}, lastRun: "", lastError: "", lastDurationMs: 0 };
    if (!existsSync(this.statePath)) return empty;
    try { return { ...empty, ...(JSON.parse(readFileSync(this.statePath, "utf8")) as SyncState) }; } catch { return empty; }
  }

  /** Atomic write of sync.json. A failure is recorded in lastError, never thrown: the replica must outlive a full disk. */
  saveState() {
    try {
      const tmp = this.statePath + ".tmp";
      writeFileSync(tmp, JSON.stringify(this.state, null, 2));
      renameSync(tmp, this.statePath);
    } catch (e) {
      this.state.lastError = `state: ${(e as Error).message}`.slice(0, 500);
    }
  }

  markFtsBuilt(table: string) {
    const st = this.state.tables[table] ?? { cursor: null, rows: 0, lastPull: "" };
    st.ftsBuilt = true;
    this.state.tables[table] = st;
    this.saveState();
  }

  async connect(): Promise<lancedb.Connection> {
    if (!this.db) this.db = await lancedb.connect(this.dir);
    return this.db;
  }

  async table(spec: TableSpec): Promise<lancedb.Table> {
    const db = await this.connect();
    const names = await db.tableNames();
    if (names.includes(spec.name)) return db.openTable(spec.name);
    return db.createEmptyTable(spec.name, spec.schema);
  }

  /** Pull everything new for every table. Returns rows upserted. Serialised: a second call joins the running one. */
  syncNow(): Promise<number> {
    if (this.running) return this.running;
    this.running = this.doSync().finally(() => { this.running = null; });
    return this.running;
  }

  private async doSync(): Promise<number> {
    const t0 = Date.now();
    let total = 0;
    try {
      for (const spec of TABLES) total += await this.syncTable(spec);
      this.state.lastError = "";
    } catch (e) {
      this.state.lastError = `${(e as Error).message}`.slice(0, 500);
    }
    this.state.lastRun = new Date().toISOString();
    this.state.lastDurationMs = Date.now() - t0;
    this.saveState();
    return total;
  }

  private async syncTable(spec: TableSpec): Promise<number> {
    const tbl = await this.table(spec);
    const st: TableState = this.state.tables[spec.name] ?? { cursor: null, rows: 0, lastPull: "" };
    let pulled = 0;
    // PocketBase ids are random, so a row committed after a page was read, in
    // the same millisecond as the cursor and with a lower id, would sort before
    // the cursor and be missed forever. Each run therefore starts a little
    // before the saved cursor; the re-pulled rows are upserts, so the only
    // cost is a few extra rows per run. Paging within the run stays exact.
    let cursor: Cursor | null = st.cursor ? rewind(st.cursor, REWIND_MS) : null;
    for (;;) {
      const page = await this.pb.pageAfter(spec.name, spec.stamp, cursor);
      if (page.items.length === 0) break;
      const rows = page.items.map(r => project(spec, r));
      await tbl.mergeInsert("id").whenMatchedUpdateAll().whenNotMatchedInsertAll().execute(makeArrowTable(rows, { schema: spec.schema }));
      const last = page.items[page.items.length - 1];
      cursor = { stamp: String(last[spec.stamp] ?? last.created), id: last.id };
      // never move the saved cursor backwards (the rewind window re-reads old rows)
      if (!st.cursor || cmpCursor(cursor, st.cursor) > 0) st.cursor = cursor;
      st.rows += rows.length;
      st.lastPull = new Date().toISOString();
      pulled += rows.length;
      this.state.tables[spec.name] = st;
      this.saveState();
      if (page.items.length < 1000) break;
    }
    this.state.tables[spec.name] = st;
    if (spec.fts && pulled > 0) await this.ensureFts(spec, tbl, st);
    if (pulled > 0) await this.maybeOptimize(spec, tbl);
    return pulled;
  }

  /**
   * Every merge page is a new Lance version and a new fragment. Left alone,
   * a busy day is thousands of versions and tens of gigabytes (26GB for one
   * events table on 2026-09-10). optimize() compacts fragments, folds new rows
   * into the FTS index, and prunes versions older than PRUNE_AFTER_MS — far
   * shorter than the 7-day default, which is meant for readers that pin old
   * versions; nothing here does.
   */
  private async maybeOptimize(spec: TableSpec, tbl: lancedb.Table) {
    const due = this.optimizeDueFor.get(spec.name) ?? 0;
    if (Date.now() < due) return;
    this.optimizeDueFor.set(spec.name, Date.now() + OPTIMIZE_EVERY_MS);
    await tbl.optimize({ cleanupOlderThan: new Date(Date.now() - PRUNE_AFTER_MS) });
  }

  /** Build the FTS index once the table has rows; afterwards fold new rows in with optimize(). */
  private async ensureFts(spec: TableSpec, tbl: lancedb.Table, st: TableState) {
    if (!spec.fts) return;
    if (!st.ftsBuilt) {
      if ((await tbl.countRows()) === 0) return;
      await tbl.createIndex(spec.fts, { config: Index.fts(), replace: true });
      st.ftsBuilt = true;
    }
    // new rows are folded into the index by maybeOptimize()
  }

  /** Follow the store: sync now, then again on every live message or every `intervalMs`. */
  async follow(intervalMs: number, log: (s: string) => void) {
    // realtime wake-up; the poll below is the safety net when the stream drops
    (async () => {
      while (!this.stopped) {
        try {
          await this.pb.live("structor/live", () => this.wake?.());
        } catch (e) {
          log(`live feed ${this.target.name}: ${(e as Error).message}`);
        }
        await Bun.sleep(5000);
      }
    })();
    let lastLogged = "";
    while (!this.stopped) {
      try {
        const n = await this.syncNow();
        if (n > 0) log(`${this.target.name}: +${n} rows`);
        else if (this.state.lastError && this.state.lastError !== lastLogged) log(`${this.target.name}: ${this.state.lastError}`);
        lastLogged = this.state.lastError;
      } catch (e) {
        // doSync already catches; this guards the loop itself so the process never dies here
        log(`${this.target.name}: ${(e as Error).message}`);
      }
      await new Promise<void>(res => {
        const t = setTimeout(res, intervalMs);
        this.wake = () => { clearTimeout(t); this.wake = null; res(); };
      });
      // let a burst of ingests coalesce into one pull
      await Bun.sleep(1500);
    }
  }

  stop() { this.stopped = true; this.wake?.(); }

  /** Rows on the PocketBase side minus rows here, per table — the lag the admin shows. */
  async lag(): Promise<Record<string, { remote: number; local: number }>> {
    const out: Record<string, { remote: number; local: number }> = {};
    for (const spec of TABLES) {
      const [remote, local] = await Promise.all([
        this.pb.count(spec.name).catch(() => -1),
        this.table(spec).then(t => t.countRows()).catch(() => -1),
      ]);
      out[spec.name] = { remote, local };
    }
    return out;
  }
}

const REWIND_MS = 2000;
export const OPTIMIZE_EVERY_MS = 5 * 60_000;
export const PRUNE_AFTER_MS = 60 * 60_000;

/** Cursor `ms` earlier than c, id cleared so every row at that stamp qualifies. PocketBase stamps look like "2026-09-09 15:00:00.100Z". */
export function rewind(c: Cursor, ms: number): Cursor {
  const t = Date.parse(c.stamp.replace(" ", "T"));
  if (Number.isNaN(t)) return c;
  const iso = new Date(t - ms).toISOString(); // 2026-09-09T15:00:00.100Z
  return { stamp: iso.replace("T", " "), id: "" };
}

/** Order two cursors the way PocketBase sorts them: by stamp, then id. */
export function cmpCursor(a: Cursor, b: Cursor): number {
  if (a.stamp !== b.stamp) return a.stamp < b.stamp ? -1 : 1;
  return a.id === b.id ? 0 : a.id < b.id ? -1 : 1;
}

/** Shape a PocketBase record into the table's fixed schema (unknown keys dropped, JSON stringified). */
export function project(spec: TableSpec, r: PBRecord): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const f of spec.schema.fields) {
    const v = r[f.name];
    if (f.type instanceof Utf8) out[f.name] = v == null ? "" : typeof v === "string" ? v : JSON.stringify(v);
    else if (f.type instanceof Float64) out[f.name] = typeof v === "number" ? v : Number(v ?? 0) || 0;
    else if (f.type instanceof Bool) out[f.name] = Boolean(v);
    else out[f.name] = v ?? null;
  }
  return out;
}
